debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/SnapshotMySQLUtility.java

```diff
@@ -0,0 +1,600 @@
+                                   BufferedBlockingConsumer<SourceRecord> bufferedRecordQueue)
+            throws SQLException {
+        if (!tablesLocked) {
+            throw new IllegalStateException("No locked tables to buffer");
+        }
+
+        // todo: verify that all tables to be buffered are locked?
+
+        final long bufferingStartTs = context.clock().currentTimeInMillis();
+        final MySqlSchema schema = context.dbSchema();
+
+        if (minimalBlocking) {
+            if (lockedTables != null) {
+                // We could not acquire a global read lock and instead had to obtain individual table-level read locks
+                // using 'FLUSH TABLE <tableName> WITH READ LOCK'. However, if we were to do this, the 'UNLOCK TABLES'
+                // would implicitly commit our active transaction, and this would break our consistent snapshot logic.
+                // Therefore, we cannot unlock the tables here!
+                // https://dev.mysql.com/doc/refman/5.7/en/flush.html
+                logger.info("tables were locked explicitly, but to get a consistent snapshot we cannot "
+                    + "release the locks until we've read all tables.");
+            } else {
+                // We are doing minimal blocking via a global read lock, so we should release the global read lock now.
+                // All subsequent SELECT should still use the MVCC snapshot obtained when we started our transaction
+                // (since we started it "...with consistent snapshot"). So, since we're only doing very simple SELECT
+                // without WHERE predicates, we can release the lock now ...
+                logger.info("releasing global read lock to enable MySQL writes");
+                unlockTables();
+            }
+        }
+
+        // Dump all of the tables and generate source records ...
+        logger.info("scanning contents of {} tables while still in transaction", tableIdsToBuffer.size());
+        metrics.setTableCount(tableIdsToBuffer.size());
+
+        long startScan = context.clock().currentTimeInMillis();
+        AtomicLong totalRowCount = new AtomicLong();
+        int counter = 0;
+        int completedCounter = 0;
+        long largeTableCount = context.rowCountForLargeTable();
+        Iterator<TableId> tableIdIter = tableIdsToBuffer.iterator();
+        while (tableIdIter.hasNext()) {
+            TableId tableId = tableIdIter.next();
+
+            // Obtain a record maker for this table, which knows about the schema ...
+            RecordMakers.RecordsForTable recordMaker = context.makeRecord().forTable(tableId, null, bufferedRecordQueue);
+            if (recordMaker != null) {
+
+                // Switch to the table's database ...
+                sql.set("USE " + quote(tableId.catalog()) + ";");
+                mysql.execute(sql.get());
+
+                AtomicLong numRows = new AtomicLong(-1);
```

Comment:
E.g. this block, couldn't we extract something like `statementFactory = buildStatementFactory()` which would then encapsulate the entire logic for building the SF?